providers/src/airflow/providers/google/cloud/operators/pubsub.py

```diff
@@ -774,32 +782,61 @@ def __init__(
         self.ack_messages = ack_messages
         self.messages_callback = messages_callback
         self.impersonation_chain = impersonation_chain
+        self.deferrable = deferrable
+        self.poll_interval = poll_interval
 
     def execute(self, context: Context) -> list:
-        hook = PubSubHook(
-            gcp_conn_id=self.gcp_conn_id,
-            impersonation_chain=self.impersonation_chain,
-        )
-
-        pulled_messages = hook.pull(
-            project_id=self.project_id,
-            subscription=self.subscription,
-            max_messages=self.max_messages,
-            return_immediately=True,
-        )
-
-        handle_messages = self.messages_callback or self._default_message_callback
-
-        ret = handle_messages(pulled_messages, context)
+        if self.deferrable:
+            self.defer(
+                trigger=PubsubPullTrigger(
+                    subscription=self.subscription,
+                    project_id=self.project_id,
+                    max_messages=self.max_messages,
+                    ack_messages=self.ack_messages,
+                    gcp_conn_id=self.gcp_conn_id,
+                    poke_interval=self.poll_interval,
+                    impersonation_chain=self.impersonation_chain,
+                ),
+                method_name=GOOGLE_DEFAULT_DEFERRABLE_METHOD_NAME,
+            )
+        else:
+            hook = PubSubHook(
+                gcp_conn_id=self.gcp_conn_id,
+                impersonation_chain=self.impersonation_chain,
+            )
 
-        if pulled_messages and self.ack_messages:
-            hook.acknowledge(
+            pulled_messages = hook.pull(
                 project_id=self.project_id,
                 subscription=self.subscription,
-                messages=pulled_messages,
+                max_messages=self.max_messages,
+                return_immediately=True,
             )
 
-        return ret
+            handle_messages = self.messages_callback or self._default_message_callback
+
+            ret = handle_messages(pulled_messages, context)
+
+            if pulled_messages and self.ack_messages:
+                hook.acknowledge(
+                    project_id=self.project_id,
+                    subscription=self.subscription,
+                    messages=pulled_messages,
+                )
+
+            return ret
+
+    def execute_complete(self, context: Context, event: dict[str, Any]):
+        """If messages_callback is provided, execute it; otherwise, return immediately with trigger event message."""
+        if event["status"] == "success":
+            self.log.info("Sensor pulls messages: %s", event["message"])
+            if self.messages_callback:
+                received_messages = self._convert_to_received_messages(event["message"])
+                _return_value = self.messages_callback(received_messages, context)
```

Comment:
Should we use the `_default_message_callback` here the same way as it was used in the non-deferrable mode?

Reply:
Thank you!